pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -181,33 +188,49 @@ public void seal()
     }
 
     // Create bitmaps from inverted index buffers and serialize them to file
-    try (DataOutputStream offsetDataStream = new DataOutputStream(
-        new BufferedOutputStream(new FileOutputStream(_invertedIndexFile)));
-        FileOutputStream bitmapFileStream = new FileOutputStream(_invertedIndexFile);
-        DataOutputStream bitmapDataStream = new DataOutputStream(new BufferedOutputStream(bitmapFileStream))) {
-      int bitmapOffset = (_cardinality + 1) * Integer.BYTES;
-      offsetDataStream.writeInt(bitmapOffset);
-      bitmapFileStream.getChannel().position(bitmapOffset);
-
+    ByteBuffer offsetBuffer = null;
+    ByteBuffer bitmapBuffer = null;
+    try (FileChannel channel = new RandomAccessFile(_invertedIndexFile, "rw").getChannel()) {
```

Comment:
Where exactly is the potential benefit coming from?

Reply:
Two places.

- We found that serializing to `ByteBuffer` could be up to 20x faster than to `DataOutput`, which was why the API was added. 
- If the bitmaps being materialised prior to serialisation below get reasonably large, using `RoaringBitmapWriter` improves the efficiency of construction. This is because when you call `RoaringBitmap.add`, it does a binary search in the high 16 bits of the value to figure out where to put the low 16 bits, but `RoaringBitmapWriter` assumes all adds are appends and optimises for it.

Reply:
From the code, I feel the benefit of serializing to `ByteBuffer` is to avoid the reverseBytes caused by BE to LE. How could it cause 20x performance difference? (Or the output stream is not buffered maybe?)

Reply:
I had a hard time believing it when this came up, but did enough profiling to figure out what was going on. The difference is _up to_ 20x, and it depends on the composition of the bitmap: `DataOutput` penalises dense bitmaps more than others. `DataOutput` has to do things like manually extract bytes from `int`, `long` etc. and this adds up to a lot compared to a simple move instruction. Because of endianness, you get away without reversing the bytes, and with `ByteBuffer` various bounds checks get eliminated. You can see the difference even with a pre-sized `ByteArrayOutputStream`, so it's not a question of unbuffered IO.